pkg/telnetlib/connection.go

```diff
@@ -0,0 +1,297 @@
+	serverOpts map[byte]bool
+	clientOpts map[byte]bool
+
+	// dataRW is the data buffer. It is written to by the FSM and read from by the data handler
+	dataRW        io.ReadWriter
+	cmdBuffer     bytes.Buffer
+	fsm           *telnetFSM
+	handlerWriter io.WriteCloser
+	// cmdHandler is the command handler for the telnet server
+	// it is a callback function when receiving commands issued by the telnet client
+	// it is wrapped by the cmdHandlerWrapper
+	cmdHandler CmdHandlerFunc
+
+	// dataHandler is the data handler of the telnet server
+	// it is a call back function when receiving data from the telnet client
+	// it is wrapped by the dataHandlerWrapper
+	dataHandler DataHandlerFunc
+
+	// used in the dataHandlerWrapper to notify that the telnet connection is closed
+	dataHandlerCloseCh chan chan struct{}
+	// used in the dataHandlerWrapper to notify that data has been writeen to the dataRW buffer
+	dataWrittenCh chan bool
+
+	// callBack function for receiving an option during negotiation
+	optionCallback func(byte, byte)
+
+	// connWriteDoneCh closes the write loop when the telnet connection is closed
+	connWriteDoneCh chan chan struct{}
+	// negotiationDone notifies that telnet negotiation has been done
+	negotiationDone chan struct{}
+
+	closedMutex sync.Mutex
+	closed      bool
+}
+
+// Safely read/write concurrently to the data Buffer
+// databuffer is written to by the FSM and it is read from by the dataHandler
+type dataReadWriter struct {
+	dataBuffer bytes.Buffer
+	dataMux    *sync.Mutex
+}
+
+func (drw *dataReadWriter) Read(p []byte) (int, error) {
+	drw.dataMux.Lock()
+	defer drw.dataMux.Unlock()
+	return drw.dataBuffer.Read(p)
+}
+
+func (drw *dataReadWriter) Write(p []byte) (int, error) {
+	drw.dataMux.Lock()
+	defer drw.dataMux.Unlock()
+	return drw.dataBuffer.Write(p)
+}
+
+type connectionWriter struct {
+	ch chan []byte
+}
+
+func (cw *connectionWriter) Write(b []byte) (int, error) {
+	if cw.ch != nil {
+		cw.ch <- b
+	}
+	return len(b), nil
+}
+
+func (cw *connectionWriter) Close() error {
+	close(cw.ch)
+	cw.ch = nil
+	return nil
+}
+
+func newTelnetConn(opts connOpts) *TelnetConn {
+	tc := &TelnetConn{
+		conn:               opts.conn,
+		writeCh:            make(chan []byte),
+		unackedServerOpts:  make(map[byte]bool),
+		unackedClientOpts:  make(map[byte]bool),
+		cmdHandler:         opts.cmdHandler,
+		dataHandler:        opts.dataHandler,
+		dataHandlerCloseCh: make(chan chan struct{}),
+		dataWrittenCh:      make(chan bool),
+		serverOpts:         opts.serverOpts,
+		clientOpts:         opts.clientOpts,
+		optionCallback:     opts.optCallback,
+		connWriteDoneCh:    make(chan chan struct{}),
+		negotiationDone:    make(chan struct{}),
+		closed:             false,
+	}
+	if tc.optionCallback == nil {
+		tc.optionCallback = tc.handleOptionCommand
+	}
+	tc.handlerWriter = &connectionWriter{
+		ch: tc.writeCh,
+	}
+	tc.dataRW = &dataReadWriter{
+		dataMux: new(sync.Mutex),
+	}
+	fsm := opts.fsm
+	fsm.tc = tc
```

Comment:
Since a telnet connection is only instantiated through Accept(), then we are guaranteed that fsm is not nil